### PR TITLE
[WIP] The big Composed Component refactor

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

Getting rid of Composed Component and moving functionality to Position Component.

Fixes # (issue number, if there is one)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] This branch is based on `develop`
- [x] This PR is targeted to merge into `develop` (not `master`)
- [x] I have formatted my code with `flutter format`
- [ ] I have made corresponding changes to the documentation
- [x] I have added examples for new features in `doc/examples`
- [ ] The continuous integration (CI) is passing

### CODE REVIEW

**Comment by spydon** [comment]
We could add all the .gitignore stuff to the root gitignore so that new contributors don't have to.

File: doc/examples/audiopool/lib/main.dart [hunk 1]
```diff
@@ -11,17 +11,16 @@ import 'package:flutter/gestures.dart';
...
   runApp(game.widget);
-
-  final TapGestureRecognizer taps = TapGestureRecognizer()
-    ..onTapDown = (_) => game.tap();
-  Flame.util.addGestureRecognizer(taps);
 }
 
-TextConfig regular = TextConfig(color: BasicPalette.white.color);
-AudioPool pool = AudioPool('laser.mp3');
+final regular = TextConfig(color: BasicPalette.white.color);
```
Comment (spydon) [inline doc/examples/audiopool/lib/main.dart:21]: Why remove the types?

File: doc/examples/audiopool/lib/main.dart [hunk 2]
```diff
@@ -11,17 +11,16 @@ import 'package:flutter/gestures.dart';
...
+final pool = AudioPool('laser.mp3');
+final black = BasicPalette.black.paint;
```
Comment (renancaraujo) [inline doc/examples/audiopool/lib/main.dart:23]: Keep PRs like this as dry as possible.

Comment (renancaraujo) [inline doc/examples/audiopool/lib/main.dart:23]: Keep PRs like this as dry as possible.

File: doc/examples/box2d/contact_callbacks/lib/boundaries.dart [hunk 3]
```diff
@@ -32,8 +32,8 @@ class Wall extends BodyComponent {
 
   @override
   void renderPolygon(Canvas canvas, List<Offset> coordinates) {
-    Offset start = coordinates[0];
-    Offset end = coordinates[1];
+    final start = coordinates[0];
+    final end = coordinates[1];
```
Comment (renancaraujo) [inline doc/examples/box2d/contact_callbacks/lib/boundaries.dart:36]: ditto

File: doc/examples/README.md [hunk 4]
```diff
@@ -5,3 +5,11 @@ This folder contains a collection of more specific examples using flame, special
...
 
 After that, you should be able to `flutter run` just fine.
+
+In order to simplify the understanding of each examples, they are all missing the following folders:
+
+```
+android, ios, web, macos, test
+```
+
+So that we can focus on the `lib` and (if applicable) `assets` folder. Please add that to the gitignore of your new example.
```
Comment (renancaraujo) [inline doc/examples/README.md:15]: Is it possible to split these changes to a different PR? 

I am afraid this onde became unnecessarily big and it is doing a lot more than what is in the title/description.

File: lib/components/flare_component.dart [hunk 5]
```diff
@@ -31,7 +31,7 @@ class FlareComponent extends PositionComponent {
 
   @override
   void render(Canvas canvas) {
-    prepareCanvas(canvas);
+    super.render(canvas);
     _flareAnimation.render(canvas, x: 0, y: 0);
   }
```
Comment (renancaraujo) [inline lib/components/flare_component.dart:36]: hehehe let this one dieeeee

File: lib/components/sprite_component.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import 'dart:ui' show Canvas, Paint;
+
+import '../sprite.dart';
+import 'component.dart';
+import 'position_component.dart';
+
+/// A [PositionComponent] that renders a single [Sprite] at the designated position,
+/// scaled to have the designated size and rotated to the designated angle.
+///
+/// This is the most commonly used child of [Component].
```
Comment (renancaraujo) [inline lib/components/sprite_component.dart:10]: > This is the most commonly used child of [Component].

Is that still true? I mean, PositionComponent is a "child" of Component.

Comment (renancaraujo) [inline lib/components/sprite_component.dart:10]: Also, flutters docs rearely use "parent/child" terminology for class inheritance, instead, you can use "subclass".

File: lib/components/sprite_component.dart [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  SpriteComponent.fromSprite(double width, double height, this.sprite) {
+    this.width = width;
+    this.height = height;
+  }
+
+  @override
+  void render(Canvas canvas) {
+    super.render(canvas);
+    sprite.render(canvas,
+        width: width, height: height, overridePaint: overridePaint);
```
Comment (renancaraujo) [inline lib/components/sprite_component.dart:32]: Trailing comma

File: lib/components/position_component.dart [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+  @mustCallSuper
+  @override
+  void render(Canvas canvas) {
+    prepareCanvas(canvas);
+    canvas.save();
+    _children.forEach((comp) => _renderChild(canvas, comp));
+    canvas.restore();
+  }
+
+  void _renderChild(Canvas canvas, Component c) {
```
Comment (vladyslav-dotsenko) [inline lib/components/position_component.dart:171]: Just wondering if that would handle composed priority?
e.g. you set custom priority for a parent and all it's children are rendered with correct order.

File: doc/examples/effects/lib/main.dart [hunk 9]
```diff
@@ -0,0 +1,117 @@
+import 'package:flutter/material.dart';
```
Comment (erickzanardo) [inline doc/examples/effects/lib/main.dart:1]: This is the default boilerplate from `flutter create .`, any reasons to have added this?

File: lib/components/flare_component.dart (hunk 5 above)
Comment (erickzanardo) [reply lib/components/flare_component.dart:36]: don't talk like that :(

File: doc/examples/audiopool/lib/main.dart (hunk 2 above)
Comment (luanpotter) [reply doc/examples/audiopool/lib/main.dart:23]: I can split, I just cannot stand linting issues while I was looking at each example code to see if it was working.

File: lib/components/position_component.dart (hunk 8 above)
Comment (luanpotter) [reply lib/components/position_component.dart:171]: they are within each level, not across the tree. so all siblings that are within the same parent when we do _children.forEach it goes by priority because of the ordered_set

File: doc/examples/effects/lib/main.dart (hunk 9 above)
Comment (luanpotter) [reply doc/examples/effects/lib/main.dart:1]: no, didn't noticed

File: lib/components/position_component.dart [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+      renderDebugMode(canvas);
+    }
+  }
+
+  void addEffect(PositionComponentEffect effect) {
+    _effects.add(effect..component = this);
+  }
+
+  void propagateToChildren<T extends PositionComponent>(
+      void Function(T, Rect) handler) {
```
Comment (luanpotter) [inline lib/components/position_component.dart:139]: trailing comma!

**Comment by luanpotter** [comment]
Closing in lieu of other PRs, in particular https://github.com/flame-engine/flame/pull/440
